### PR TITLE
Add auto-update notifications and confirmation

### LINKED ISSUE #1768: Prompt for auto-update
Changes the auto-update to not be automatic.

* If a new version is found during the application runtime, the user is asked _once_ in a prompt, e.g. "New version available. Would you like to update"?
* Adds auto-update section to e.g. Preferences page, which displays a button to update if a version is available.

Fixes #1701, #1689 and the groundwork for #1601.

### PR DESCRIPTION
**Pull request by Nokel81**
This is mostly a cherry pick of #1831 with small changes to make it work with other changes made so far in 4.x.

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Final PR in the auto-updater saga.

Closes #1768

### CODE REVIEW

**Review by jimmyesquire — DISMISSED** [review]

File: scripts/integration.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    find ~/.config/Logs/Lens -type f -name *.log -delete
+  ;;
+  win)
+    find %APPDATA%/Logs/Lens -type f -name *.log -delete
+  ;;
+esac
+
+make build-extension-types build-extensions
+yarn build:$1
+DEBUG=true yarn integration-runner
```
Comment (jim-docker) [inline scripts/integration.sh]: what is `integration-runner`?

File: src/renderer/themes/kontena-dark.json [hunk 2]
```diff
@@ -0,0 +1,112 @@
+{
```
Comment (jim-docker) [inline src/renderer/themes/kontena-dark.json]: renamed from lens-dark.json(?)

Comment (Nokel81) [reply src/renderer/themes/kontena-dark.json]: Oops thanks for the catch

File: scripts/integration.sh (hunk 1 above)
Comment (Nokel81) [reply scripts/integration.sh]: See `package.json`

Comment (jim-docker) [reply scripts/integration.sh]: I don't see it. What am I looking for?

Comment (Nokel81) [reply scripts/integration.sh]: Oh I see, I forgot to cherry pick that in, good catch.

File: src/main/app-updater.ts [hunk 3]
```diff
@@ -1,20 +1,158 @@
...
+  autoUpdater.autoDownload = false;
+  autoUpdater.autoInstallOnAppQuit = false;
+
+  autoUpdater
+    .on("update-available", async (args: UpdateInfo) => {
+      try {
+        const releaseDate = moment(args.releaseDate);
+        const body = `Version ${args.version} was released on ${releaseDate.format("dddd, MMMM Do, yyyy")}.`;
+
+        windowManager.sendToView({
```
Comment (jakolehm) [inline src/main/app-updater.ts]: Why not `broadcastMessage()`?

File: src/renderer/components/notifications/notifications.store.tsx [hunk 4]
```diff
@@ -26,6 +62,29 @@ export class NotificationsStore {
 
   protected autoHideTimers = new Map<NotificationId, number>();
 
+  registerIpcListener(): void {
+    logger.info(`[NOTIFICATION-STORE] start to listen for notifications requests from main`);
+    ipcRenderer.on(NotificationChannelAdd, (event, model: MainNotification) => {
```
Comment (jakolehm) [inline src/renderer/components/notifications/notifications.store.tsx]: Why not `subscribeToBroadcast()`?

File: src/renderer/components/notifications/notifications.store.tsx [hunk 5]
```diff
@@ -18,6 +22,38 @@ export interface Notification {
...
+    return null;
+  }
+
+  return (
+    <>
+      <br />
+      <div className="ButtonPannel flex row align-right box grow">
+        {buttons.map(({ backchannel, ...props }) => (
+          <Button key={backchannel} {...props} onClick={() => {
+            ipcRenderer.send(backchannel);
```
Comment (jakolehm) [inline src/renderer/components/notifications/notifications.store.tsx]: Does this mean that it's only possible to send notifications from `main`?  (`broadcastMessage()` probably would solve this)

Comment (Nokel81) [reply src/renderer/components/notifications/notifications.store.tsx]: I don't understand, `renderer` already has a way to display notifications.

File: src/renderer/components/notifications/notifications.store.tsx (hunk 4 above)
Comment (Nokel81) [reply src/renderer/components/notifications/notifications.store.tsx]: So either each `iframe` has its own `NotificationStore` (or that used to be the case in 3.X) and we don't want notifications from main to be duplicated or there is only one `NotifcationStore` for all of the `renderer` and this works (without do all the extra work of sending to the other frames). 

Lastly, I don't think that `main` should have to care about `NotificationId` to avoid duplicate rendered notifications.

File: src/renderer/components/notifications/notifications.store.tsx (hunk 5 above)
Comment (jakolehm) [reply src/renderer/components/notifications/notifications.store.tsx]: But that's different way, right? Would it make sense to have just one way to trigger notifications (might be out of scope of this PR).

Comment (Nokel81) [reply src/renderer/components/notifications/notifications.store.tsx]: I think that is out of scope for this PR. Yes it is a different way. They call `Notification.(info|error|success)(...)`

File: src/main/app-updater.ts [hunk 6]
```diff
@@ -1,20 +1,124 @@
...
-    autoUpdater.logger = logger;
+        resolve();
+      })
+      .on(noChannel, () => {
+        logger.info("[UPDATE CHECKER]: User chose not to update");
+        cleanupChannels();
+        resolve();
+      });
+
+    windowManager.sendToView({
```
Comment (jakolehm) [inline src/main/app-updater.ts]: I feel this is tight coupling of main & renderer & ui. Could we just broadcast a message that an update is available (with update info). Based on that message renderer could show a notification. On notification action renderer could send a message back to main. Meaning: main should not need to know how update information is drawn, it just provides the information and channel for triggering an update process.

Comment (Nokel81) [reply src/main/app-updater.ts]: Do you mean the `status` ad `action` fields below? Because the rest of that data is quite required. I don't think that it is tight coupling give that `main` the source of a notification should decided what is displayed on it.

Though if you are talking about the "sendToView" vs "broadcast". I don't think we should broadcast because then `main` has to worry about setting an ID so that the notification is only displayed once (which I don't think `main` should need to care about)

Comment (jakolehm) [reply src/main/app-updater.ts]: > Do you mean the `status` ad `action` fields below? Because the rest of that data is quite required. I don't think that it is tight coupling give that `main` the source of a notification should decided what is displayed on it.

Not really those specific fields but this logic in general. Why can't `main` just broadcast:
```ts
broadcastMessage("app:update-available", updateInfo);
```

and a component in renderer could listen this and act accordingly? Why `main` needs to know about buttons, labels etc.. it should be `renderer` decision to draw information.


> Though if you are talking about the "sendToView" vs "broadcast". I don't think we should broadcast because then `main` has to worry about setting an ID so that the notification is only displayed once (which I don't think `main` should need to care about)

IMHO `main` should not care who is listening these kind of messages as long as these messages stay within core application. Even if we send this to just a view how can `main` be sure that it's displayed only once?

Comment (nevalla) [reply src/main/app-updater.ts]: > Meaning: main should not need to know how update information is drawn, it just provides the information and channel for triggering an update process.

IMHO this sounds reasonable. Main just send information and renderer decides how the notification looks like. Ideally we should have only one way to trigger those notifications.

File: scripts/integration.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  linux)
+    find ~/.config/Logs/Lens -type f -name *.log -delete
+  ;;
+  win)
+    find %APPDATA%/Logs/Lens -type f -name *.log -delete
+  ;;
+esac
+
+make build-extension-types build-extensions
+yarn build:$1
```
Comment (jakolehm) [inline scripts/integration.sh]: This feels a bit odd.. could you explain why we refactored away from top-level make targets?

File: src/common/ipc.ts [hunk 8]
```diff
@@ -29,16 +29,21 @@ export function broadcastMessage(channel: string, ...args: any[]) {
...
+  views.forEach(async webContent => {
     const type = webContent.getType();
 
     logger.silly(`[IPC]: broadcasting "${channel}" to ${type}=${webContent.id}`, { args });
     webContent.send(channel, ...args);
-    getSubFrames().then((frames) => {
-      frames.map((frameInfo) => {
+
+    try {
+      const frames = await getSubFrames();
```
Comment (jakolehm) [inline src/common/ipc.ts]: This actually changes how message ordering goes (just a bit but it might break expected behaviour).. Please revert unless there is a real reason why this is done (I tried to make similar change in #1957 but it did break some of the app functionality).

File: src/renderer/components/notifications/notifications.store.tsx (hunk 5 above)
Comment (jakolehm) [reply src/renderer/components/notifications/notifications.store.tsx]: Related to my other review comment (https://github.com/lensapp/lens/pull/1941#r565863019)): could we just use `Notification` component to show a notification in `renderer`? If no -> could we enhance it so that it can handle these kind of use cases. We really should have just one way to trigger notifications (note: `Notification` is also exposed via extension API).

File: scripts/integration.sh (hunk 7 above)
Comment (Nokel81) [reply scripts/integration.sh]: Because I don't think that makefile supports what I am doing with lines 17-31. Specifically checking the return.

File: src/common/ipc.ts (hunk 8 above)
Comment (Nokel81) [reply src/common/ipc.ts]: How does this change the messaging order? I didn't make line 32 a `for-loop` so these will still happen asynchronously (though actually concurrently because JS is single-treaded)

Comment (jakolehm) [reply src/common/ipc.ts]: Based on discussions & testing it seems that this does not have side-effects. So probably fine, although I still don't see how this is related to this specific PR.

File: scripts/integration.sh (hunk 7 above)
Comment (Nokel81) [reply scripts/integration.sh]: However, some searching has found that I can do it

File: src/common/ipc.ts (hunk 8 above)
Comment (Nokel81) [reply src/common/ipc.ts]: I don't remember either. Will pull out and put into own PR

**Comment by Nokel81** [comment]
Here is a verification image of the notification styling:

Screen Shot 2021-01-29 at 2 26 52 PM

Note: macOS only has "yes" and "no" whereas linux and windows will have "yes, now", "yes, on quit", and "no"

File: Makefile [hunk 9]
```diff
@@ -38,18 +38,21 @@ test: binaries/client
 
 .PHONY: integration-linux
 integration-linux: build-extension-types build-extensions
+	-find ~/.config/Logs/Lens -type f -name *.log -delete
```
Comment (jakolehm) [inline Makefile]: These are probably debug leftovers?

File: src/main/app-updater.ts [hunk 10]
```diff
@@ -1,20 +1,71 @@
...
-    autoUpdater.logger = logger;
+  autoUpdater.logger = logger;
+  autoUpdater.autoDownload = false;
+  autoUpdater.autoInstallOnAppQuit = false;
+
+  autoUpdater
+    .on("update-available", (args: UpdateInfo) => {
+      try {
+        // use a UUID so that this back-channel is harder to discover
+        const backchannel = uuid.v4();
```
Comment (jakolehm) [inline src/main/app-updater.ts]: Why do we need "hidden" backchannel? Could backchannel be just  something like `auto-update:${version}`.

File: Makefile (hunk 9 above)
Comment (Nokel81) [reply Makefile]: I sort of wanted them to help with debugging in the future, but I'll open a separate PR for that after 4.1

File: src/main/app-updater.ts (hunk 10 above)
Comment (Nokel81) [reply src/main/app-updater.ts]: This is to try and discourage malicious extensions from spamming this backchannel

**Comment by jakolehm** [comment]
@Nokel81 please resolve conflicts.

File: src/main/app-updater.ts (hunk 10 above)
Comment (jakolehm) [reply src/main/app-updater.ts]: How could extensions spam this? They don't have access to this IPC.

**Comment by Nokel81** [comment]
@jakolehm done

File: package.json [hunk 11]
```diff
@@ -212,6 +212,7 @@
     "mobx-observable-history": "^1.0.3",
     "mobx-react": "^6.2.2",
     "mock-fs": "^4.12.0",
+    "moment": "^2.26.0",
```
Comment (aleksfront) [inline package.json]: Why we're moved out `moment` from `devDependencies`?

File: src/common/ipc/update-available/index.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+export const UpdateAvailableChannel = "update-available";
+export const AutoUpdateLogPrefix = "[UPDATE-CHECKER]";
+
+/**
+ * [<back-channel>, <update-info>]
+ */
+export type UpdateAvailableFromMain = [string, UpdateInfo];
+
+export function areArgsUpdateAvailableFromMain(args: unknown[]): args is UpdateAvailableFromMain {
```
Comment (aleksfront) [inline src/common/ipc/update-available/index.ts:11]: What exactly these typing mean? `args is UpdateAvailableFromMain` and at the same time `args: unknown[]`? Is it possible to determine argument types instead of `unknown`?

File: src/common/ipc/update-available/index.ts [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  if (args.length !== 2) {
```
Comment (aleksfront) [inline src/common/ipc/update-available/index.ts:12]: Probably need to export to separate function because there's no clear description of this and following conditions. For example, why we return `false` if there's less or more 2 arguments? Separating into function with meaningful name will show the condition intent.

File: src/renderer/components/button/button-panel.tsx [hunk 14]
```diff
@@ -0,0 +1,13 @@
+import "./button-panel.scss";
+import React from "react";
+
+export function ButtonPannel(props: React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement>) {
```
Comment (aleksfront) [inline src/renderer/components/button/button-panel.tsx]: If I understand it right, this `ButtonPanel` shows buttons with a `margin-right` added to last button. Is it reasonable to create separate component for such a small purpose? I suggest to add flex styles in place where it's needed without creating `ButtonPanel`

File: src/renderer/components/button/button.scss [hunk 15]
```diff
@@ -45,6 +51,7 @@
   &.outlined {
     color: inherit;
     background: transparent;
+    border: 1px solid;
```
Comment (aleksfront) [inline src/renderer/components/button/button.scss]: Please check if apperance of button inside `<InfoPanel/>` stays same because of this change. Search shows that there at least 1 button that using `outlined` style in our codebase.

File: src/common/ipc/ipc.ts [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+  source.on(channel, handler);
+}
+
+/**
+ * Adds a listener to `source` that checks to verify the arguments before calling the handler.
+ * @param channel The channel to be listened on
+ * @param listener The function for the channel to be called if the args of the correct type
+ * @param verifier The function to be called to verify that the args are the correct type
+ */
+export function onCorrect<
```
Comment (aleksfront) [inline src/common/ipc/ipc.ts]: Do this function is necessary to be added in current PR? I'm worried that such addition would make auto-update issue quite complex.

**Review by aleksfront — DISMISSED** [review]

File: package.json (hunk 11 above)
Comment (Nokel81) [reply package.json]: I think I initially did so because we were formatting a date for the notification. However, looking now I see that we use it all over the place in `renderer` so we need it at runtime and thus should be part of `deps`

File: src/common/ipc/update-available/index.ts (hunk 12 above)
Comment (Nokel81) [reply src/common/ipc/update-available/index.ts:11]: The is what Typescipt calls a "user defined type-guard" (https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards).

Namely, it is a function that is used to validate that the inputs are a certain type and thus narrow them (if the function returns `true`).

File: src/common/ipc/update-available/index.ts (hunk 13 above)
Comment (Nokel81) [reply src/common/ipc/update-available/index.ts:12]: I completely disagree the function name is not meaningful and signature explains exactly what this function is trying to do.

IPC calls can happen from anywhere so the handler that is called gets some array of elements (namely `unknown[]`). I use `unknown` here instead of `any` because of the difference of the coercion rules. Plus typescipt now recommends to use `unknown` where you don't know the types.

Finally, `UpdateAvailableFromMain` is by defenition a tuple (or an array of length 2). So if the arguments passed to the fuction are not of length 2 we know that they are invalid and thus `!(args is UpdateAvailableFromMain)`.

File: src/main/app-updater.ts (hunk 10 above)
Comment (Nokel81) [reply src/main/app-updater.ts]: Do extensions not have access to `ipcRenderer`?

File: src/renderer/components/button/button-panel.tsx (hunk 14 above)
Comment (Nokel81) [reply src/renderer/components/button/button-panel.tsx]: The `margin-right` is added to all buttons except that last one.

I was under the impression that such small components were exactly what React's design methodology was about, which is why I created one.

File: src/main/app-updater.ts (hunk 10 above)
Comment (jakolehm) [reply src/main/app-updater.ts]: Lens is not exposing any `ipc` stuff to extensions. When (if) we expose `ipc` to extensions it most likely will be prefixed so that extensions cannot cause trouble to Lens internal communication (like stores).

Comment (Nokel81) [reply src/main/app-updater.ts]: Okay I can change over to an incrementing channel

File: src/common/ipc/ipc.ts (hunk 16 above)
Comment (Nokel81) [reply src/common/ipc/ipc.ts]: How does this add undue complexity. I find the use of this function much less complex because it helps enforce typing through the IPC channel (and logs when they differ).

File: src/renderer/ipc/index.tsx [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+function UpdateAvailableHandler(event: IpcRendererEvent, ...[backchannel, updateInfo]: UpdateAvailableFromMain): void {
+  const notificationId = uuid.v4();
+
+  function sendToBackchannel(data: BackchannelArg): void {
+    notificationsStore.remove(notificationId);
+    console.log("sending to backchanel", { backchannel, data });
+    ipcRenderer.send(backchannel, data);
+  }
+
+  function renderYesButtons() {
```
Comment (ixrock) [inline src/renderer/ipc/index.tsx]: Could be better to put `RenderYesButtons` name and use as `<ReactComponent/>` inside notification?

Comment (Nokel81) [reply src/renderer/ipc/index.tsx]: Probably, good idea

File: src/renderer/ipc/index.tsx [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+      id: notificationId,
+      onClose() {
+        sendToBackchannel({ doUpdate: false });
+      }
+    }
+  );
+}
+
+export function registerIpcHandlers() {
+  onCorrect(ipcRenderer, UpdateAvailableChannel, UpdateAvailableHandler, areArgsUpdateAvailableFromMain);
```
Comment (ixrock) [inline src/renderer/ipc/index.tsx]: I would more expect to see something:

```ts
return unregisterIpcHandlers(() => onCorrect(ipcRenderer, UpdateAvailableChannel, UpdateAvailableHandler, areArgsUpdateAvailableFromMain)); // returns unsubscribe: () => void
```

File: src/common/ipc/ipc.ts [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+// https://www.electronjs.org/docs/api/ipc-main
+// https://www.electronjs.org/docs/api/ipc-renderer
+
+import { ipcMain, ipcRenderer, webContents, remote } from "electron";
+import { toJS } from "mobx";
+import { EventEmitter } from "ws";
+import logger from "../../main/logger";
+import { ClusterFrameInfo, clusterFrameMap } from "../cluster-frames";
+
+const subFramesChannel = "ipc:get-sub-frames";
```
Comment (ixrock) [inline src/common/ipc/ipc.ts:10]: Shouldn't we prefix `subFramesChannel` with corresponding frame-id?

File: src/common/ipc/ipc.ts [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+  source: EM,
+  channel: string | symbol,
+  listener: L,
+  verifier: (args: unknown[]) => args is T
+): void {
+  function handler(event: HandlerEvent<EM>, ...args: unknown[]): void {
+    if (verifier(args)) {
+      source.removeListener(channel, handler); // remove immediately
+
+      Promise.resolve(listener(event, ...args)) // might return a promise
```
Comment (ixrock) [inline src/common/ipc/ipc.ts]: If `listener(event, ...args)` will throw would it be caught in `.catch()`?

File: src/renderer/ipc/index.tsx (hunk 18 above)
Comment (Nokel81) [reply src/renderer/ipc/index.tsx]: Why do you expect that the handlers will be removed? These are supposed to be present for the entire lifetime of the `renderer`.

File: src/common/ipc/ipc.ts (hunk 19 above)
Comment (Nokel81) [reply src/common/ipc/ipc.ts:10]: Did I copy this over wrong (let me check)?

File: src/common/ipc/ipc.ts (hunk 20 above)
Comment (Nokel81) [reply src/common/ipc/ipc.ts]: No, (though I thought it did). Will fix, thanks.

File: src/renderer/ipc/index.tsx (hunk 18 above)
Comment (ixrock) [reply src/renderer/ipc/index.tsx]: Btw, there is not big correlation with `onCorrect` arguments and their names.

Would be better/more readable if `onCorrect` could accept object-map (and this declaration actually easier to extend later when needed, especially critical when you export something like to extensions-api).

```ts
onCorrect({
  source: ipcRenderer,
  channel: UpdateAvailableChannel,
  listener: UpdateAvailableHandler,
  verifier: areArgsUpdateAvailableFromMain
})
```

Also when I open a file with `areArgsUpdateAvailableFromMain` there is also nothing about that it's some kinda `verifier` argument. Maybe better to name files with a suffix `*.verifier.ts`? Wdyt?

Comment (Nokel81) [reply src/renderer/ipc/index.tsx]: Do you mean a verifier function? I figured that it was obvious from the signature (specifically its declared return value). But I can move it to a different function.

As for taking an object instead of positional args. Sure.

File: src/renderer/components/button/button.scss [hunk 21]
```diff
@@ -21,10 +21,16 @@
   &.primary {
     background: $buttonPrimaryBackground;
   }
+
   &.accent {
     background: $buttonAccentBackground;
   }
 
+  &.light {
+    background-color: white;
```
Comment (ixrock) [inline src/renderer/components/button/button.scss]: Is it works good with dark/light themes in Lens?

Comment (Nokel81) [reply src/renderer/components/button/button.scss]: Will check, this was copied directly from the changes Alexei made for the 3.6.10 release. Will verify.

File: src/renderer/ipc/index.tsx [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+import { Button, ButtonPannel } from "../components/button";
+import { isMac } from "../../common/vars";
+import * as uuid from "uuid";
+
+function UpdateAvailableHandler(event: IpcRendererEvent, ...[backchannel, updateInfo]: UpdateAvailableFromMain): void {
+  const notificationId = uuid.v4();
+
+  function sendToBackchannel(data: BackchannelArg): void {
+    notificationsStore.remove(notificationId);
+    console.log("sending to backchanel", { backchannel, data });
```
Comment (ixrock) [inline src/renderer/ipc/index.tsx]: Forgotten `console.log`?

File: src/common/ipc/ipc.ts [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+import { ipcMain, ipcRenderer, webContents, remote } from "electron";
+import { toJS } from "mobx";
+import { EventEmitter } from "ws";
+import logger from "../../main/logger";
+import { ClusterFrameInfo, clusterFrameMap } from "../cluster-frames";
+
+const subFramesChannel = "ipc:get-sub-frames";
+
+export type HandlerEvent<EM extends EventEmitter> = Parameters<Parameters<EM["on"]>[1]>[0];
+export type EventListener<E extends EventEmitter, T extends any[]> = (event: HandlerEvent<E>, ...args: T) => any;
```
Comment (ixrock) [inline src/common/ipc/ipc.ts]: Seems not used `type` anywhere.

File: src/common/ipc/ipc.ts (hunk 19 above)
Comment (Nokel81) [reply src/common/ipc/ipc.ts:10]: This is copied directly from `master` and I haven't touched the code there.

File: src/renderer/components/button/button.scss (hunk 15 above)
Comment (Nokel81) [reply src/renderer/components/button/button.scss]: I removed it because it increases the height by 2px.

File: src/renderer/components/button/button.scss (hunk 21 above)
Comment (Nokel81) [reply src/renderer/components/button/button.scss]: It looks sort of okay. I think I'll change it to "invert".

File: src/renderer/components/button/button.scss (hunk 15 above)
Comment (Nokel81) [reply src/renderer/components/button/button.scss]: So the old `outlined` styling only worked if the button wasn't also styled the same as the background (namely you couldn't have a `primary outlined` button on `primary` notification. That is why I initially had this change.

File: src/renderer/components/button/button-panel.tsx (hunk 14 above)
Comment (aleksfront) [reply src/renderer/components/button/button-panel.tsx]: For creating gaps (`margin-right`) between buttons you can use following classes on the parent element
```
<div className="flex gaps">...</div>
```
The `.gaps` rule would add neede margins.

Please look if `ButtonPanel` component could be avoided.

Comment (Nokel81) [reply src/renderer/components/button/button-panel.tsx]: Tested and removed.

File: package.json (hunk 11 above)
Comment (jakolehm) [reply package.json]: Renderer deps are bundled with webpack. Most likely this should be under `devDependencies`.

File: src/common/ipc/ipc.ts (hunk 19 above)
Comment (jakolehm) [reply src/common/ipc/ipc.ts:10]: Why did we move `ipc` here?

File: src/common/ipc/update-available/index.ts [hunk 24]
```diff
@@ -0,0 +1,48 @@
+import { UpdateInfo } from "electron-updater";
```
Comment (jakolehm) [inline src/common/ipc/update-available/index.ts:1]: I'm not sure why we need to put these modules under `ipc/`? If we go with this route every module that uses `ipc` should end up here (which sounds wrong).

Comment (Nokel81) [reply src/common/ipc/update-available/index.ts:1]: So you would prefer having these checking function under where? I put them here because these IPC types should be "public" to both `renderer` and `main` (as the receiver and sender) and I figured that the verification functions should be beside the type definitions.

File: src/common/ipc/ipc.ts (hunk 19 above)
Comment (Nokel81) [reply src/common/ipc/ipc.ts:10]: Do you mean why was `src/common/ipc.ts` was moved to `src/common/ipc/ipc.ts`? That is because I added another module file. This doesn't change any imports because of how `index.ts` is handled.

Comment (jakolehm) [reply src/common/ipc/ipc.ts:10]: Ok, asked because it's quite hard to spot differences when whole file is moved and modified 🙂

File: src/common/ipc/update-available/index.ts (hunk 24 above)
Comment (jakolehm) [reply src/common/ipc/update-available/index.ts:1]: Ok, makes sense if this is just shared functionality.

File: src/common/ipc/ipc.ts (hunk 16 above)
Comment (jakolehm) [reply src/common/ipc/ipc.ts]: `onCorrect` sounds a bit abstract.. maybe we can improve names in future? (naming is hard)

Comment (Nokel81) [reply src/common/ipc/ipc.ts]: Sounds good about improving the names in the future. For reference I chose this name because it is sort of equivalent to `ipc.on(...)` but guarantees `correct` arguments.

File: package.json (hunk 11 above)
Comment (Nokel81) [reply package.json]: Okay will do

File: src/common/ipc/ipc.ts (hunk 19 above)
Comment (Nokel81) [reply src/common/ipc/ipc.ts:10]: Fair, should I move the argument checking functions to a different file so that this file is just a rename in the git diff?

Comment (Nokel81) [reply src/common/ipc/ipc.ts:10]: Nvm, I'll do that anyway

**Comment by Nokel81** [comment]
@aleksfront and @jakolehm please re review, I move some code around to other files so that it is easier to review.

**Review by jakolehm — APPROVED** [review]
LGTM

File: src/common/utils/delay.ts [hunk 25]
```diff
@@ -1,6 +1,8 @@
-// Create async delay for provided timeout in milliseconds
-
-export async function delay(timeoutMs = 1000) {
-  if (!timeoutMs) return;
-  await new Promise(resolve => setTimeout(resolve, timeoutMs));
+/**
```
Comment (ixrock) [inline src/common/utils/delay.ts:1]: Why this change is needed? :) 
Also I would prefer to keep `timeoutMs` as it is more clear/descriptive without consulting the docs..

**Review by ixrock — APPROVED** [review]

File: src/common/utils/delay.ts (hunk 25 above)
Comment (Nokel81) [reply src/common/utils/delay.ts:1]: I had this function as part of my code base before it go merged to master. I can revert the `timeoutMs` in a followup PR if you want.

Though side note, what editor are you using? I use vscode and it always shows the full docs (including the tsdocs) when I hover over a function.

Comment (ixrock) [reply src/common/utils/delay.ts:1]: I use `WebStorm` and this is what I see when typing:

image

Comment (ixrock) [reply src/common/utils/delay.ts:1]: Hovering works, but usually I don't use it (if I need to read docs I'd better navigate to the source):

image

**Comment by hansiemithun** [comment]
This is not working anymore, the Lens is automatically updated on reopen. 

**Environment:** mac
